### ISSUE TITLE
[New Arch] RTL to LTR of vise versa Layout Direction Change Requires App Termination to Apply

### ISSUE THREAD

**Issue by a-eid** [facebook/react-native#48311]
# Description 

When toggling RTL layout direction dynamically using I18nManager.forceRTL(true) or switching between languages in the system settings, the Yoga layout engine does not immediately apply the updated layout direction. Even with a forced app restart (e.g., via RNRestart), the changes are not respected.

A complete termination and relaunch of the app is required for the changes to take effect. This creates significant friction for users and developers aiming to support dynamic language and layout changes for right-to-left (RTL) languages.

# Report

- [x] I have searched [existing issues](https://github.com/facebook/yoga/issues) and this is not a duplicate

# Issues and Steps to Reproduce

Toggle RTL programmatically:

```js
I18nManager.forceRTL(true);
I18nManager.allowRTL(true);
RNRestart.Restart();
```

Alternatively, change the device language to an RTL language (e.g., Arabic ) in the app settings. ( assuming you added such functionality )

Observe that the layout direction does not change immediately.

Terminate the app completely and relaunch it manually. Only then does the layout respect the RTL direction.

# Expected Behavior
 the layout direction should change.
 
# Actual Behavior
 the layout direction does not change.

[related react-native Issue ](https://github.com/facebook/react-native/issues/45661)

**Comment by NickGerleman** [comment]
Moved to RN repo since this is implementation issue with RN iOS New arch instead of Yoga layer.

**Comment by a-eid** [comment]
@NickGerleman unfortunately the RN issue is locked and I don't think there is any plans to fix this issue any time soon. 
it's blocking many RTL apps from migrating to the new arch. 

I thought there might be an issue with how yoga handles RTL & LTR with the new arch..

**Comment by Linuhusainnk** [comment]
Any update on this?

**Comment by MehadND** [comment]
Any update?

**Comment by a-eid** [comment]
@blakef are there any plans to fix this issue ? most apps are moving to the new arch, we're unable to because of this issue unfortunately. I'd appreciate if you could share some updates.

**Comment by blakef** [comment]
AFAIK there isn't work planned to fix this. PRs are always welcome. 

Cc @cipolleschi, might know more.

**Comment by a-eid** [comment]
With libraries targeting only the new arch, eg: reanimated v4, unistyles v3 & shadowlist, I think this is very essential to fix.

**Comment by blakef** [comment]
No, it isn't.

**Comment by zaidqureshi95** [comment]
anyone fixed the issue?

**Comment by Makhdoom-Sharif** [comment]
I am experiencing the same issue. Has anyone found a solution?

**Comment by Abdulazeez-98** [comment]
I faced the same problem when I updated to RN 0.76 (expo-52). One solution that worked for me is to specify the 'direction' property of the <NavigationContainer /> component (from '@react-navigation/native') which wraps the whole app (set it to either 'ltr' or 'rtl').

**Comment by a-eid** [comment]
@Abdulazeez-98 I thought this would only control the stack navigation direction, does it also control the layout direction ?

**Comment by Abdulazeez-98** [comment]
> [@Abdulazeez-98](https://github.com/Abdulazeez-98) I thought this would only control the stack navigation direction, does it also control the layout direction ?

In my case, it solved the issue for the whole app. I applied the 'ltr' layout to the main navigator (which is a bottom tab navigator with 4 nested stack navigators). I think nested Views inherits this property.

**Comment by sabuhiteymurov** [comment]
I think this is a major issue which should not be ignored by the core team. We are unable to migrate to new architecture because of this issue.

**Comment by a-eid** [comment]
Sadly, it doesn't look like Meta is prioritizing this fix. However, @jamonholmgren and @mazenchami from Infinite Red have shown interest in tackling it, so hopefully, it’ll be resolved soon.

**Comment by andremawad** [comment]
The below patch has fixed the issue for me: 

```
diff --git a/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm b/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
index 7b3ce59..3906df7 100644
--- a/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
+++ b/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
@@ -143,6 +143,7 @@ - (RCTSurfaceView *)view
 
   if (!_view) {
     _view = [[RCTSurfaceView alloc] initWithSurface:(RCTSurface *)self];
+    [self _updateLayoutContext];
     _touchHandler = [RCTSurfaceTouchHandler new];
     [_touchHandler attachToView:_view];
   }
```

**Comment by a-eid** [comment]
@andremawad yeah, thanks hopefully @chrsmys PR will be merged soon.

**Comment by mazenchami** [comment]
it looks like 0.78 was just released but after Meta validates it against their `main` it might out with 0.78.1

**Comment by cortinico** [comment]
> it looks like 0.78 was just released but after Meta validates it against their `main` it might out with 0.78.1

Yup 0.78 was just out, but given the patch is a one-liner, this is easy to integrate in other previous versions as well

**Comment by cipolleschi** [comment]
make sure to open [pick requests](https://github.com/reactwg/react-native-releases/issues/new?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen&template=pick_request_form.yml). Please, one pick request per each RN version! 🙏

**Comment by cortinico** [comment]
Closing as this was solved with https://github.com/facebook/react-native/pull/49455

**Comment by SyedZawwarAhmed** [comment]
> [self _updateLayoutContext];

This does fix the issue.

**Comment by SyedZawwarAhmed** [comment]
> The below patch has fixed the issue for me:
> 
> ```
> diff --git a/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm b/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
> index 7b3ce59..3906df7 100644
> --- a/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
> +++ b/node_modules/react-native/React/Fabric/Surface/RCTFabricSurface.mm
> @@ -143,6 +143,7 @@ - (RCTSurfaceView *)view
>  
>    if (!_view) {
>      _view = [[RCTSurfaceView alloc] initWithSurface:(RCTSurface *)self];
> +    [self _updateLayoutContext];
>      _touchHandler = [RCTSurfaceTouchHandler new];
>      [_touchHandler attachToView:_view];
>    }
> ```

This does fix the issue.

**Comment by a-eid** [comment]
@cortinico  
it's unfortunate this is not 100% fixed. it's gonna be a huge issue with RN 82 since we can't turn the new arch off.

**Comment by cortinico** [comment]
> [@cortinico](https://github.com/cortinico) it's unfortunate this is not 100% fixed. it's gonna be a huge issue with RN 82 since we can't turn the new arch off.

Please create a new issue + provide repro then

**Comment by ashish0990** [comment]
I have an Question Is this possible to change direction of language without restart the applicaiton

#reactNative #LTR-RTL

**Comment by a-eid** [comment]
@ashish0990 No it's not ... @chrsmys mentioned that it's possible in his PR .. but I don't think there were any progress on it.
